tool/planet/membership.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/planet/lib/monitoring"
+
+	"github.com/gravitational/trace"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// startSerfReconciler creates a control loop that periodically attempts to
+// reconcile serf cluster.
+func startSerfReconciler(ctx context.Context, serfConfig *serf.Config) {
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			k8sPeers, err := getK8sPeers()
+			if err != nil {
+				log.WithError(err).Warn("Failed to query kubernetes peers.")
+				break
+			}
+
+			serfPeers, err := getSerfPeers(serfConfig)
+			if err != nil {
+				log.WithError(err).Warn("Failed to query serf peers.")
+				break
+			}
+
+			if !shouldReconcileSerf(k8sPeers, serfPeers) {
+				break
+			}
+
+			log.WithField("k8s-peers", k8sPeers).
+				WithField("serf-peers", serfPeers).
+				Info("Reconciling serf cluster.")
+
+			numJoined, err := reconcileSerfCluster(serfConfig, k8sPeers)
+			if err != nil {
+				log.WithError(err).Warn("Failed to reconcile serf cluster.")
+				break
+			}
+			log.WithField("num-joined", numJoined).Info("Reconciled serf cluster.")
+
+		case <-ctx.Done():
+			log.Debug("Stopping serf reconciler")
+			return
+		}
+	}
+}
+
+// getK8sPeers returns advertised IP addresses of all nodes in the kubernetes
+// cluster.
+func getK8sPeers() (peers []string, err error) {
+	client, err := monitoring.GetKubeClient()
+	if err != nil {
+		return peers, trace.Wrap(err, "failed to get kubernetes clientset")
+	}
+	nodes, err := client.CoreV1().Nodes().List(metav1.ListOptions{
+		LabelSelector: advertiseIPKey,
```

Comment:
What's this label selector for?

Reply:
This is to ensure we are only selecting gravity nodes. Not sure if it is guaranteed that all nodes in the kubernetes cluster is part of the gravity cluster.

Reply:
Yes, we do not support adding non-Gravity-managed nodes to the cluster, so I think this filter is redundant.